rfcs/2020-09-16-3875-kubernetes-metrics-api.md

```diff
@@ -0,0 +1,262 @@
+
+  - Custom Metrics API
+
+    - Defined [here](https://github.com/kubernetes/metrics/blob/bd98ade7905d9e394dfa9fcf1eee4e2a3966e9a9/pkg/apis/custom_metrics/v1beta2/types.go).
+    - Design [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/custom-metrics-api.md).
+
+    The API specifies the shape of the objects suitable to handle a generic
+    metric value on an arbitrary Kubernetes API object.
+    It's all meta and such, but essentially the types defined are
+    `MetricIdentifier`, `MetricListOptions`, `MetricValue` and
+    `MetricValueList`.
+    You get the idea - generic ways to describe generic metrics about something.
+
+    We can, for root-scoped and namespaced objects, do:
+
+    - `GetForObject` - obtain a `MetricValue` by a Kubernetes API Object
+      group and name, and a metric name and selector;
+    - `GetForObjects` - obtain a `MetricValueList` by a Kubernetes API Object
+      group and selector, and a metric name and selector.
+
+    See the
+    [generated client](https://github.com/kubernetes/metrics/blob/bd98ade7905d9e394dfa9fcf1eee4e2a3966e9a9/pkg/client/custom_metrics/versioned_client.go)
+
+  - External Metrics API
+
+    - Defined [here](https://github.com/kubernetes/metrics/blob/bd98ade7905d9e394dfa9fcf1eee4e2a3966e9a9/pkg/apis/external_metrics/v1beta1/types.go).
+    - Design [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/external-metrics-api.md).
+
+    This API defines just two types - `ExternalMetricValue` and
+    `ExternalMetricValueList`.
+
+    These are just metrics, identified by the name and and a list of labels
+    (kv pairs of string) - no relation to the Kubernetes API resource like with
+    at Custom Metrics API.
+
+    All we can do with this API is get an `ExternalMetricValueList` by the
+    namespace, metric name and metrics selector.
+
+  The implementation of these are typically vendor-specific, and expose
+  information about the cloud that Kubernetes runs atop and/or surrounding
+  infrastructure.
+
+Check out a list of
+[implementations](https://github.com/kubernetes/metrics/blob/bd98ade7905d9e394dfa9fcf1eee4e2a3966e9a9/IMPLEMENTATIONS.md)
+of the Metrics API to better understand the idea on how it all works.
+
+> We're currently focusing on integrating with just the Resource Metrics API but
+> it's important to understand the distinction.
+
+## Should we use it
+
+Overall, it looks like Metrics API is designed to allow external monitoring
+tools to route data in such a way that it's available in a "native" fashion
+to the Horizontal Pod Autoscaler and other Kubernetes components that rely on
+this data. This means that while we _can_ integrate with the Metrics API
+directly, it might not be the best way, and going with an alternative - like
+integrating with those external metrics aggregation system via the Prometheus
+scraping protocol- might be a better approach.
```

Comment:
Can you clarify what this means? I didn't see any previous mention of an external metrics aggregation system.

Reply:
I have a section on this at the metrics overview at https://github.com/timberio/vector/pull/4098